### ISSUE TITLE
3.0.1 IndexDB bug wipes out sessions

### ISSUE THREAD

**Issue by Urhixidur** [sienori/Tab-Session-Manager#222]
I restarted Firefox for an update, and TSM was no longer able to access its sessions. I followed the instructions and about:config dom.indexedDB.enabled was true, "deleting" the about:profile root directory's storage folder did not help, so I uninstalled 3.0.1 and installed 2.1.1. No sessions at all. Everything is lost. I'm pissed.

**Comment by sienori** [comment]
I apologize for any inconvenience.

> I uninstalled 3.0.1 and installed 2.1.1

If you delete the add-on from about: addons, the data will be deleted.
Do not delete the add-on, please install overwrite.
I added this note to the [wiki](https://github.com/sienori/Tab-Session-Manager/wiki/IndexedDB-Error).

**Comment by sienori** [comment]
> I followed the instructions and about:config dom.indexedDB.enabled was true, "deleting" the about:profile root directory's storage folder did not help

Do you see the indexedDB error even after deleting the storage folder and restarting firefox?
![indexedDB error](https://user-images.githubusercontent.com/29514424/37827655-b2092abc-2edb-11e8-8b0a-b6813b687563.png)

**Comment by Urhixidur** [comment]
Install-overwrite was not an option (clicking the version button had no effect). And yes, the problem was still there after renaming the storage folder.

**Comment by letsprogress** [comment]
Thanks, sienori, for clarifying that re-installing 2.1.1 over 3.0.1 would return access to all my TSM sessions! What a relief.

Unclear is next step in progressing to 3.0.1 without the IndexedDB problem: 
1) should the "sessions" folder be deleted from the profile a) when 2.1.1 is installed and then install 3.0.1 over 2.1.1, b) after re-installing 3.0.1 over 2.1.1? 
2) how to access all the old sessions after the re-install of 3.0.1: do they automatically re-appear, or do they need to be imported (and how/where to import them).

I could probably trial-and-error this question with profile backups available, but I'd rather have your detailed counsel to do it right the first time. Thanks for all your work on this important extension!

**Comment by crystal90** [comment]
I followed instructions to revert back to the previous version and still lost alll my saved sessions. WTF

**Comment by BladeRunner7** [comment]
Ditto for me. Followed the steps word for word, and then reinstalled 2.1.1. Again no "install-overwrite" option. None of my original 32 tabs are there. This is very upsetting. Not just "inconvient".

**Comment by avaxcina** [comment]
How to recover sessions after going through this "IndexedDB Error"? This is so inconvient...

**Comment by sienori** [comment]
@Urhixidur 
> Install-overwrite was not an option

"install-overwrite" means the following:
Install [version 2.1.1](https://addons.mozilla.org/en-US/firefox/addon/tab-session-manager/versions/#version-2.1.1) from "Add to Firefox" button without uninstalling version 3.1.1.
![add to firefox](https://user-images.githubusercontent.com/29514424/38348096-ccfb6b06-38da-11e8-8e8b-4a92f45977e9.png)

> yes, the problem was still there after renaming the storage folder.

Hmm... Just to be sure, try deleting the storage folder completely.

**Comment by sienori** [comment]
@letsprogress 
1. If there is no IndexedDB error, you do not need to delete the "storage" folder.
2. Older sessions are loaded automatically after installing 3.0.1.

**Comment by sienori** [comment]
@crystal90 @BladeRunner7 
Is this what you mean?
1. First, you installed version 3.0.1, but IndexedDB error appeared
2. You installed version 2.1.1
3. However, old sessions are gone

There are several possible causes.
* You deleted version 3.0.1 before installing version 2.1.1
  Once you delete an add-on, data may be lost.
* Sessions disappeared due to a storage API bug when installing version 2.1.1
  Before version 2.1.1, a bug that the session disappears has been reported.
  At the time of starting version 2.1.1, the session might have disappeared unluckily.
* Sessions have already been migrated to IndexedDB
  When version 3.0.1 starts up successfully without error, sessions are migrated to IndexedDB.
  Sessions migrated to IndexedDB are not visible in version 2.1.1.

If the session has already disappeared, there is no way to restore it. I'm sorry.
I recommend follow the [wiki](https://github.com/sienori/Tab-Session-Manager/wiki/IndexedDB-Error) instructions and using latest version.

**Comment by sienori** [comment]
@avaxcina 
Please see the [wiki](https://github.com/sienori/Tab-Session-Manager/wiki/IndexedDB-Error).

**Comment by zankai2** [comment]
So I just nearly lost all my sessions by following the instructions by mistake.  They don't apply to crashes!

Firefox crashed, and upon restarting I got the Indexed DB error.

The solution is to close Firefox.  
Rename profile/storage to storage2.
Restart Firefox.
You will see Tab Session Manager shows no sessions.
Now rename /profile/storage2 to storage.  It may say file in use error.  Keep trying.
Once renamed, your sessions should reappear.

**Comment by Urhixidur** [comment]
Fresh boot of machines today. IndexedDB bug strikes (using version 3.0.1). dom.indexedDB.enabled is true. Close Firefox, rename storage folder, start Firefox. IndexedDB failure is still present. I will now overwrite with 2.1.1 ---should I restore storage first?

**Comment by john3voltas** [comment]
@sienori 
Sorry to bump in on the conversation. I just experienced the exact same problem as @crystal90 and @BladeRunner7 .
Thank god I had a very recent backup of my sessions because I couldn't revive my sessions...
In order to avoid any future issues, maybe you could possibly add a step in your article on how to over come this issue so that people back their sessions in advance of doing anything or they risk losing all their windows/tabs.
Just a thought...
Cheers

**Comment by BladeRunner7** [comment]
Thanks for your input. Is there an ideal way to save the files so that they are readable  by the updated version?
I had tried, but did something wrong, as they were not being read or recognized by Firefox and/or the TSM software.

Thanks,

RQ2

Get Outlook for Android<https://aka.ms/ghei36>


________________________________

**Comment by john3voltas** [comment]
@BladeRunner7 
You mean the backup files being human readable or the backup files being "restorable" on a newer TSM version?
On the first case, I don't think there is a way for those files being human readable. At all.
On the second case, my backup was very recent but was from a version prior to 3.x and I was able to restore it on 3.x.
Not sure if a very old backup would work on 3.x, though. Maybe only @sienori can clear that up for us.
Cheers

**Comment by jimmyjudas** [comment]
I found the instructions on the wiki a bit confusing. Was I meant to downgrade the version or follow steps 1-4, or both? I tried following steps 1-4 but it didn't work. Instead, I just imported (Settings > Sessions > Import Sessions) one of the auto-backups (Settings > Settings > Backup). It's not perfect though - that backup is from the last time Firefox started, so if you'd created any new sessions between Firefox starting and crashing then you'd have lost these.

Still, I think the wiki could be made a bit more clear. Thanks!

**Comment by Urhixidur** [comment]
@AngryGami that may have been your case, but it certainly wasn't mine.

**Comment by grahamperrin** [comment]
If you haven't already seen it: <https://bugzilla.mozilla.org/show_bug.cgi?id=944918#c32> and subsequent commentary. 

tl;dr

- removing the entire storage directory, or any database, may be overkill
- if you're lucky, simply targeting some types of hidden file will allow indexedDB to work as expected. 

## Side note re: possible underlying causes

Where KDE is used, I have a hunch <https://www.reddit.com/r/kde/comments/8hl0ql/-/dylk7kz/?context=1> that offending files may be more likely to reappear if e.g. a recent release of Dolphin is used to browse the directory hierarchy … or something like that. Unconfirmed.

**Comment by Helstar74** [comment]
Hello, I had this problem today (wish i could know beforehand... you should have sent a warning/notification or something). Everything is gone and i don't know how to restore the sessions. I tried overwriting 2.1.1, deleting storage folder (of course i made a backup of it), putting again the last version, and all the combinations possible, nothing. I'm really mad now...

**Comment by grahamperrin** [comment]
@Helstar74 hi. I received, via subscription, the comment that you posted yesterday, Sunday 2018-06-03 ~16:54 (local time, I guess – UK here, Monday morning now).

From the disappearance of your comment, I assume that you resolved (or worked around) the problem in your case. True? If so: can you please summarise the resolution/workaround? 

Also, let us know which operating system (e.g. Mac OS X) and desktop environment or file manager (e.g. Finder) you use.

Thanks. 

----

If you still have the backup: please zip it – to prevent hidden changes – then set aside the `.zip` file. Its **untouched content** might include something that will significantly help to minimise problems for other users of things (such as Tab Session Manager) that use IndexedDB.

**Comment by RobertAb1** [comment]
@sienori 
[Bug 1423917](https://bugzilla.mozilla.org/show_bug.cgi?id=1423917) was fixed.

**Comment by RobertAb1** [comment]
https://bugzilla.mozilla.org/show_bug.cgi?id=944918#c69
